python/dgl/backend/chainer/tensor.py

```diff
@@ -0,0 +1,536 @@
+        self.retain_outputs((0,))
+
+        return out,
+
+    def forward_gpu(self, inputs):
+        X, = inputs
+        feat_shape = X.shape[1:]
+
+        X_nd = _zerocopy_to_dgl_ndarray(X)
+        out = cupy.empty((self.out_size,) + feat_shape, dtype=X.dtype)
+        out_nd = _zerocopy_to_dgl_ndarray(out)
+
+        K.copy_reduce(
+            self.reducer, self.G, self.target, X_nd, out_nd,
+            self.X_rows[0], self.out_rows[0])
+
+        self.retain_outputs((0,))
+
+        return out,
+
+    def backward_cpu(self, inputs, grad_outputs):
+        X, = inputs
+        out, = self.output_data
+        grad_out, = grad_outputs
+
+        X_nd = _zerocopy_to_dgl_ndarray(X)
+        out_nd = _zerocopy_to_dgl_ndarray(out)
+        grad_out_nd = _zerocopy_to_dgl_ndarray(grad_out)
+
+        grad_X_nd = nd.empty(X.shape, X.dtype, nd.cpu())
+        K.backward_copy_reduce(
+            self.reducer, self.G, self.target, X_nd, out_nd,
+            grad_out_nd, grad_X_nd,
+            self.X_rows[1], self.out_rows[1])
+        # TODO (ZEROCOPY_BETWEEN_NUMPY_AND_DGL)
+        grad_X = _zerocopy_from_dgl_ndarray(grad_X_nd)
+
+        return grad_X,
+
+    def backward_gpu(self, inputs, grad_outputs):
+        X, = inputs
+        out, = self.output_data
+        grad_out, = grad_outputs
+        grad_out = cupy.ascontiguousarray(grad_out)
+
+        X_nd = _zerocopy_to_dgl_ndarray(X)
+        out_nd = _zerocopy_to_dgl_ndarray(out)
+        grad_out_nd = _zerocopy_to_dgl_ndarray(grad_out)
+
+        grad_X = cupy.empty(X.shape, dtype=X.dtype)
+        grad_X_nd = _zerocopy_to_dgl_ndarray(grad_X)
+        K.backward_copy_reduce(
+            self.reducer, self.G, self.target, X_nd, out_nd,
+            grad_out_nd, grad_X_nd,
+            self.X_rows[1], self.out_rows[1])
+
+        return grad_X,
+
+
+# pylint: disable=invalid-name
+def copy_reduce(reducer, G, target, X, out_size, X_rows, out_rows):
+    func = CopyReduce(reducer, G, target, out_size, X_rows, out_rows)
+    return func(X)
+
+
+def _reduce_grad(grad, shape):
+    """Reduce gradient on the broadcast dimension
+
+    If there is broadcast in forward pass, gradients need to be reduced on
+    broadcast dimension. This function checks the input tensor shape and
+    gradient shape and perform the reduction.
+
+    Parameters
+    ----------
+    grad: numpy.ndarray or cupy.ndarray
+        Gradient tensor
+    shape: tuple
+        Shape of input tensor
+
+    Returns
+    -------
+    Tensor
+    """
+    grad_shape = grad.shape[1:]
+    in_shape = shape[1:]
+    if in_shape == grad_shape:
+        # no need to reduce
+        return grad
+    num_to_squeeze = len(grad_shape) - len(in_shape)
+    # pad in_shape
+    in_shape = (1,) * num_to_squeeze + in_shape
+    reduce_idx = np.nonzero(np.array(grad_shape) - np.array(in_shape))[0]
+    reduce_idx += 1 # skip batch dim
+    grad = grad.sum(axis=tuple(reduce_idx), keepdims=True)
+    return grad.reshape(shape)
+
+
+def sync():
```

Comment:
It looks like a unique operation just for chainer, should we incorporate this?

Reply:
You mean MXNet?